tensorflow/core/kernels/training_ops_gpu.cu.cc

```diff
@@ -284,6 +340,61 @@ struct ApplyPowerSign<GPUDevice, T> {
   }
 };
 
+template <typename T>
+struct ApplyIRpropPlus<GPUDevice, T> {
+  void operator()(const GPUDevice& d, typename TTypes<T>::Flat var,
+                  typename TTypes<T>::Flat old_grad,
+                  typename TTypes<T>::Flat delta_update,
+                  typename TTypes<T>::ConstScalar eta_minus,
+                  typename TTypes<T>::ConstScalar eta_plus,
+                  typename TTypes<T>::ConstScalar delta_min,
+                  typename TTypes<T>::ConstScalar delta_max,
+                  typename TTypes<T>::ConstScalar error,
+                  typename TTypes<T>::ConstScalar old_error,
+                  typename TTypes<T>::ConstFlat grad) {
+    const int32 N = var.size();
+
+    CudaLaunchConfig config = GetCudaLaunchConfig(N, d);
```

Comment:
Why does this optimizer need this cuda stuff while other optimizers don't?

Reply:
It is not possible to implement the update step of RProp using the GPU implementation (of the operations) of TensorFlow and Eigen-Tensor in a trivial or efficient way, therefore we use CudaLaunchConfig to determine the launch parameters (of the GPU hardware) in order to execute the CUDA kernel that computes the update step.

Reply:
Can you add this as a comment to the code, so future readers know what is going on? (specifically what eigen doesn't let you do would be very helpful)